src/core/shape_inference/include/nms_shape_inference.hpp

```diff
@@ -257,7 +245,7 @@ std::vector<TRShape> shape_infer(const NonMaxSuppression* op,
     const auto& boxes_shape = input_shapes[0];
     const auto& scores_shape = input_shapes[1];
 
-    auto output_shapes = std::vector<TRShape>{TRShape{TDim(-1), 3}};
+    auto output_shapes = std::vector<TRShape>{TRShape{TDim(dim::inf_bound), 3}};
     if (boxes_shape.rank().is_static()) {
         const auto max_out_boxes_per_class = get_input_const_data_as<TRShape, int64_t>(op, 2, ta);
         const auto max_out_class_boxes = max_out_boxes_per_class ? max_out_boxes_per_class->front() : dim::inf_bound;
```

Comment:
Shouldn't the changes be applied also for v4 nms?

Reply:
Can you explain?

Reply:
The code for shape_infer in v4 namespace is similar, but the changes in this PR are applied only to the common part.

Reply:
The v4 is different and do calculation of selected boxes only if required dimension are static so there is no need to changed.
It could be changed but inference will not work for this version in same way (not sure if any regression can pop-up). The v5 and v9 introduced more dynamic approach.

Reply:
But the case with separate input rank checks is similar and redundant, 
the calculations below `if(boxes_shape.rank().is_static())` 
have sense only when `if(scores_shape.rank().is_static())` is true, 
so it could be a single `if(boxes_shape.rank().is_static() && scores_shape.rank().is_static())`